## A "Hello World" Example of Machine Learning - Revisit

Loading the Iris dataset from scikit-learn. 

The first column represents Sepal length, the second column represents Sepal width,  the third column represents the petal length, and the fourth column the petal width of the flower samples. The classes (type of species) are already converted to integer labels where 0=Iris-Setosa, 1=Iris-Versicolor, 2=Iris-Virginica.

Here, we are using only two features: the one and fourth columns. 

In [1]:
from sklearn import datasets
import numpy as np

iris = datasets.load_iris()
#iris.data

In [2]:
X = iris.data[:, [0, 3]]

In [3]:
y = iris.target

In [4]:
print('Class labels:', np.unique(y))

Class labels: [0 1 2]


Scikit-learn algorithms support multi-class classification via the One-Versus-Rest(OvR) method. 

Splitting data into 70% training and 30% test data:

In [5]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=1, stratify=y)

In [6]:
print('Labels counts in y:', np.bincount(y))
print('Labels counts in y_train:', np.bincount(y_train))
print('Labels counts in y_test:', np.bincount(y_test))

Labels counts in y: [50 50 50]
Labels counts in y_train: [35 35 35]
Labels counts in y_test: [15 15 15]


### Standardizing the features:

In [7]:
from sklearn.preprocessing import StandardScaler

sc = StandardScaler() #center the distribution around zero (mean), with a standard deviation of 1.
sc.fit(X_train)
X_train_std = sc.transform(X_train)
X_test_std = sc.transform(X_test)

In [8]:
from sklearn.linear_model import Perceptron

ppn = Perceptron(max_iter=100, eta0=0.1, random_state=42)
ppn.fit(X_train_std, y_train)

Perceptron(alpha=0.0001, class_weight=None, early_stopping=False, eta0=0.1,
      fit_intercept=True, max_iter=100, n_iter=None, n_iter_no_change=5,
      n_jobs=None, penalty=None, random_state=42, shuffle=True, tol=None,
      validation_fraction=0.1, verbose=0, warm_start=False)

### Test the model with the hold-out test set

In [9]:
y_pred = ppn.predict(X_test_std)
print('Misclassified samples: ' + str((y_test != y_pred).sum()))

Misclassified samples: 10


In [10]:
from sklearn.metrics import accuracy_score

print('Accuracy: ' + str(accuracy_score(y_test, y_pred)))

Accuracy: 0.7777777777777778


In [11]:
X_new = [[5.4, 0.4],[5.6, 1.5], [7.7, 2.0]]
y_new = ppn.predict(X_new)
y_new

array([2, 2, 2])

### Evaluate the model using cross validation

In [12]:
from sklearn.model_selection import cross_val_score
cross_val_score(ppn, X_train_std, y_train, cv=4, scoring="accuracy")

array([0.55555556, 0.81481481, 0.96296296, 0.66666667])

2-features: sccuracy score: array([0.88888889, 0.48148148, 0.7037037 , 0.95833333])

### Exercise 1: Use all four features to train the model and use cross validaton to check if the results better? Briefly explain why. 

## perceptron using 4features 

In [13]:
from sklearn import datasets
import numpy as np
iris = datasets.load_iris()
X = iris.data[:, [0, 1, 2, 3]]

In [14]:
y = iris.target
print('Class labels:', np.unique(y))

Class labels: [0 1 2]


In [15]:
print('Labels counts in y:', np.bincount(y))
print('Labels counts in y_train:', np.bincount(y_train))
print('Labels counts in y_test:', np.bincount(y_test))

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=1, stratify=y)

Labels counts in y: [50 50 50]
Labels counts in y_train: [35 35 35]
Labels counts in y_test: [15 15 15]


In [16]:
from sklearn.preprocessing import StandardScaler

sc = StandardScaler() #center the distribution around zero (mean), with a standard deviation of 1.
sc.fit(X_train)
X_train_std = sc.transform(X_train)
X_test_std = sc.transform(X_test)

In [17]:
from sklearn.linear_model import Perceptron

ppn = Perceptron(max_iter=100, eta0=0.1, random_state=42)
ppn.fit(X_train_std, y_train)

Perceptron(alpha=0.0001, class_weight=None, early_stopping=False, eta0=0.1,
      fit_intercept=True, max_iter=100, n_iter=None, n_iter_no_change=5,
      n_jobs=None, penalty=None, random_state=42, shuffle=True, tol=None,
      validation_fraction=0.1, verbose=0, warm_start=False)

In [18]:
y_pred = ppn.predict(X_test_std)
print('Misclassified samples: ' + str((y_test != y_pred).sum()))

Misclassified samples: 1


In [19]:
from sklearn.metrics import accuracy_score
print('Accuracy: ' + str(accuracy_score(y_test, y_pred)))

Accuracy: 0.9777777777777777


In [20]:
X_new = [[5.4, 3.9, 1.7, 0.4],[5.6, 3.0, 4.5, 1.5], [7.7, 2.8, 6.7, 2.0]]
y_new = ppn.predict(X_new)
y_new

array([2, 2, 2])

In [21]:
from sklearn.model_selection import cross_val_score
cross_val_score(ppn, X_train_std, y_train, cv=4, scoring="accuracy")

array([0.92592593, 0.88888889, 0.81481481, 0.91666667])

### accuracy for 2 features = Accuracy: 0.7777777777777778
### cross_value_score for 2 features = array([0.55555556, 0.81481481, 0.96296296, 0.66666667])
________________________________________________________________________________________________
### accuracy using 4 features = Accuracy: 0.9777777777777777
### cross_value_score for 4 features = array([0.92592593, 0.88888889, 0.81481481, 0.91666667])

#### here we can see that accuracy is high with 4-features than 2-features and also cross_value_score increases for 4-features, so we can say that results are more better while using 4 features. 


### Exercise 2: Try with the scikit-learn stochastic gradient descent model instead of perceptron. Use all four features. Evaluate with cross-validation how does the model perform in terms of accuracy using both two features and four features. 

## SGD using 4 features

In [22]:
from sklearn import datasets
import numpy as np
iris = datasets.load_iris()
X = iris.data[:, [0, 1, 2, 3]]
y = iris.target
print('Class labels:', np.unique(y))
print('Labels counts in y:', np.bincount(y))
print('Labels counts in y_train:', np.bincount(y_train))
print('Labels counts in y_test:', np.bincount(y_test))
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(
X, y, test_size=0.3, random_state=1, stratify=y)
from sklearn.preprocessing import StandardScaler
sc = StandardScaler() #center the distribution around zero (mean), with a standard deviation of 1.
sc.fit(X_train)
X_train_std = sc.transform(X_train)
X_test_std = sc.transform(X_test)
from sklearn.linear_model import SGDClassifier
from sklearn.metrics import accuracy_score
sgd = SGDClassifier(random_state=42)
sgd.fit(X_train_std, y_train)
y_pred = sgd.predict(X_test_std)
print('Misclassified samples: ' + str((y_test != y_pred).sum()))
print('Accuracy: ' + str(accuracy_score(y_test,y_pred)))
X_new = [[5.4, 3.9, 1.7, 0.4],[5.6, 3.0, 4.5, 1.5], [7.7, 2.8, 6.7, 2.0]]
y_new = sgd.predict(X_new)
y_new
from sklearn.model_selection import cross_val_score
cross_val_score(sgd, X_train_std, y_train, cv=4, scoring="accuracy")

Class labels: [0 1 2]
Labels counts in y: [50 50 50]
Labels counts in y_train: [35 35 35]
Labels counts in y_test: [15 15 15]
Misclassified samples: 2
Accuracy: 0.9555555555555556


array([0.96296296, 0.88888889, 0.77777778, 0.70833333])

## SGD using 2 features

In [23]:
from sklearn import datasets
import numpy as np
iris = datasets.load_iris()
X = iris.data[:, [0, 3]]
y = iris.target
print('Class labels:', np.unique(y))
print('Labels counts in y:', np.bincount(y))
print('Labels counts in y_train:', np.bincount(y_train))
print('Labels counts in y_test:', np.bincount(y_test))
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(
X, y, test_size=0.3, random_state=1, stratify=y)
from sklearn.preprocessing import StandardScaler
sc = StandardScaler() #center the distribution around zero (mean), with a standard deviation of 1.
sc.fit(X_train)
X_train_std = sc.transform(X_train)
X_test_std = sc.transform(X_test)
from sklearn.linear_model import SGDClassifier
from sklearn.metrics import accuracy_score
sgd = SGDClassifier(random_state=42)
sgd.fit(X_train_std, y_train)
y_pred = sgd.predict(X_test_std)
print('Misclassified samples: ' + str((y_test != y_pred).sum()))
print('Accuracy: ' + str(accuracy_score(y_test,y_pred)))
X_new = [[5.4, 0.4],[5.6, 1.5], [7.7, 2.0]]
y_new = sgd.predict(X_new)
y_new
from sklearn.model_selection import cross_val_score
cross_val_score(sgd, X_train_std, y_train, cv=4, scoring="accuracy")

Class labels: [0 1 2]
Labels counts in y: [50 50 50]
Labels counts in y_train: [35 35 35]
Labels counts in y_test: [15 15 15]
Misclassified samples: 5
Accuracy: 0.8888888888888888


array([0.96296296, 0.96296296, 0.92592593, 0.83333333])

for 4 features = Accuracy: 0.9555555555555556,
for 4 features = array([0.96296296, 0.88888889, 0.77777778, 0.70833333])

for 2 features = Accuracy: 0.8888888888888888,
for 2 features = array([0.96296296, 0.96296296, 0.92592593, 0.83333333])

### even in this model accuracy is high with 4 features than 2 features. but cross_val_score is a bit high for 2 features. but sollution for one problem for a model is not same as the other model problem. when i tried with other values accuracy and cross_val_scores has been changed.
